### DISCUSSION TITLE
Getting GoError on custom code validation logic

### DISCUSSION THREAD

**Discussion by pietru2004** [pocketbase/pocketbase#3154] [Q&A] [answered]
Related https://github.com/pocketbase/pocketbase/issues/3153

I tried using this code to check if user has item
if user has no item create new item and pass it to following code
if user has item then just pass it

Code works as long as item does not exist in collection.
```
    if (money<=price){
        return c.json(400, { "message": "No enough currency.","data": {}})
    }
    let inv_record
    try {
        inv_record = $app.dao().findFirstRecordByFilter("items_inventory", "id = '"+id+"' && owner = '"+user.id+"'");
    }
    catch(err) {
        const collection = $app.dao().findCollectionByNameOrId("items_inventory")
        inv_record = new Record(collection, {
            "item": id,
            "owner": user.id,
            "count": 0
        })
    }
    inv_record.set("count", inv_record.get("count")+1)
    wallet_record.set("money", money-price)
    

    $app.dao().saveRecord(inv_record)
    $app.dao().saveRecord(wallet_record)
    
    return c.json(200, { "item": inv_record, "message": "Item bought...", "data": {}})
```

I get in logs
![obraz](https://github.com/pocketbase/pocketbase/assets/51274570/d9923d00-84a1-4aea-a992-cd381329786c)

Also unique index is used to make sure each user can only have one item
![obraz](https://github.com/pocketbase/pocketbase/assets/51274570/ad6674dd-0b3c-4b02-9c67-05cdd3cf98c1)

**Comment by ganigeorgiev** [comment, marked as answer]
The error is expected if you already have a `items_inventory` record in your collection with the matching fields pair and you are trying to create a duplicate.

I think the error in your case originate from the `try/catch` block because the filter prop probably should be `item = ...` and not `id = ...`, aka:
```js
inv_record = $app.dao().findFirstRecordByFilter("items_inventory", "item = '"+id+"' && owner = '"+user.id+"'");
```


Also keep in mind that `findFirstRecordByFilter` is not safe to use with untrusted user input.
If you can't trust `id` and `user.id` variables (there is no context in the above from they come from), then an alternative to that method could be to create your own fetch query as shown in https://pocketbase.io/docs/js-records/#custom-record-query:

```js
const record = new Record();

$app.dao().recordQuery("items_inventory")
  .andWhere($dbx.hashExp({
    "item":  id,
    "owner": user.id,
  }))
  .limit(1)
  .one(record)
```

**Reply by pietru2004** [reply]
![obraz](https://github.com/pocketbase/pocketbase/assets/51274570/0cfbdcb3-ae7f-4394-a45a-c18dccca928a)
also yes seams I missed that part with ID where it should be item... (it was the error)

**Reply by pietru2004** [reply]
also I need to ask is it save to take ID from info.authRecord and id from pathParam ?

**Reply by ganigeorgiev** [reply]
You can trust `user.id` but you can't trust the `id` variable. The reason why `findFirstRecordByFilter` could be unsafe is that an attacker could craft a custom `id` string value that will prematurely close the wrapping quote in the filter which would allow them to craft their own filter condition (_eg. they could send something like `1' || 1=1`_).
You can either ensure that the `id` variable doesn't have any single quote (eg. with `id.replace(/'/g, "")`) or use the above manual query method.

**Reply by pietru2004** [reply]
as you can see in code sample id is used to create record so I guess I gona take both - the custom querry for some, as itr looks cleaner, and use
`id = id.replace(/'/g, "")` to make it safe

**Reply by ganigeorgiev** [reply]
You don't need to manually clean the id if you use the custom query. `$dbx.hashExp` will make sure to bind the parameters to the query (see https://pocketbase.io/docs/js-database/#binding-parameters).

Both methods (the custom query and `findFirstRecordByFilter`) are safe from SQL injections.
The problem with `findFirstRecordByFilter` is not SQL injection on DB level but injection in the string filter argument.

**Reply by pietru2004** [reply]
well I still gona manualy clear that just to be safe...

**Reply by ganigeorgiev** [reply]
To minimize the footguns with `findFirstRecordByFilter()` I'll consider adding an optional second params argument that will allow safely binding user input, aka. with the next release you'll be able to do something like:

```js
$app.dao().findFirstRecordByFilter("items_inventory", "item={:item} && owner={:owner}", {
    "item": id,
    "owner": user.id,
})
```
